felix/nftables/ipsets.go

```diff
@@ -0,0 +1,513 @@
+package nftables
+
+import (
+	"context"
+	"fmt"
+	"strings"
+	"time"
+
+	"github.com/projectcalico/calico/felix/dataplane/common"
+	"github.com/projectcalico/calico/felix/deltatracker"
+	"github.com/projectcalico/calico/felix/ipsets"
+	"github.com/projectcalico/calico/libcalico-go/lib/set"
+
+	log "github.com/sirupsen/logrus"
+	"sigs.k8s.io/knftables"
+)
+
+var _ common.IPSetsDataplane = &IPSets{}
+
+// IPSets manages a whole "plane" of IP sets, i.e. all the IPv4 sets, or all the IPv6 IP sets.
+type IPSets struct {
+	IPVersionConfig *ipsets.IPVersionConfig
+
+	// setNameToAllMetadata contains an entry for each IP set that has been
+	// added by a call to AddOrReplaceIPSet (and not subsequently removed).
+	// It is *not* filtered by neededIPSetNames.
+	setNameToAllMetadata map[string]ipsets.IPSetMetadata
+
+	// setNameToProgrammedMetadata tracks the IP sets that we want to program and
+	// those that are actually in the dataplane.  It's Desired() map is the
+	// subset of setNameToAllMetadata that matches the neededIPSetNames filter.
+	// Its Dataplane() map contains all IP sets matching the IPVersionConfig
+	// that we think are in the dataplane.  This includes any temporary IP
+	// sets and IP sets that we discovered on a resync (neither of which will
+	// have entries in the Desired() map).
+	setNameToProgrammedMetadata *deltatracker.DeltaTracker[string, ipsets.IPSetMetadata]
+
+	// mainSetNameToMembers contains entries for all IP sets that are in
+	// setNameToAllMetadata along with entries for "main" (non-temporary) IP
+	// sets that we think are still in the dataplane.  It is not filtered by
+	// neededIPSetNames.  For IP sets that are in setNameToAllMetadata, the
+	// Desired() side of the tracker contains the members that we've been told
+	// about.  Otherwise, Desired() is empty.  The Dataplane() side of the
+	// tracker contains the members that are thought to be in the dataplane.
+	mainSetNameToMembers   map[string]*deltatracker.SetDeltaTracker[ipsets.IPSetMember]
+	ipSetsWithDirtyMembers set.Set[string]
+
+	sleep func(time.Duration)
+
+	resyncRequired bool
+
+	logCxt *log.Entry
+
+	// Optional filter.  When non-nil, only these IP set IDs will be rendered into the dataplane
+	// as Linux IP sets.
+	neededIPSetNames set.Set[string]
+
+	nft knftables.Interface
+}
+
+func NewIPSets(ipVersionConfig *ipsets.IPVersionConfig, nft knftables.Interface) *IPSets {
```

Comment:
This is largely a copy/paste from the existing IPSet dataplane code, but modified to render sets to nftables. Most of member / set tracking logic is exactly the same, and it's just how they are written that differs. Can probably come up with a better way to share code here, but for now this is working.